src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -216,20 +223,35 @@ object RuleUtils {
     var unhandledAppendedFiles: Seq[Path] = Nil
 
     // Get transformed plan with index data and appended files if applicable.
-    val indexPlan = plan transformDown {
+    val indexPlan = plan transformUp {
+      // Use transformUp here as currently 1 Logical Relation is allowed (pre-requisite).
+      // Otherwise, we need additional check not to process newly injected Filter-Not-In nodes.
       case baseRelation @ LogicalRelation(
             _ @HadoopFsRelation(location: PartitioningAwareFileIndex, _, _, _, _, _),
             baseOutput,
             _,
             _) =>
         val curFileSet = location.allFiles
           .map(f => FileInfo(f.getPath.toString, f.getLen, f.getModificationTime))
-        val filesAppended =
-          curFileSet.filterNot(index.allSourceFileInfos.contains).map(f => new Path(f.name))
-        // TODO: Hybrid Scan delete support.
+
+        val (filesDeleted, filesAppended) =
+          if (HyperspaceConf.hybridScanDeleteEnabled(spark) && index.hasLineageColumn(spark)) {
+            val (exist, nonExist) = curFileSet.partition(index.allSourceFileInfos.contains)
+            val filesAppended = nonExist.map(f => new Path(f.name))
+            if (exist.length < index.allSourceFileInfos.size) {
+              (index.allSourceFileInfos -- exist, filesAppended)
+            } else {
+              (Nil, filesAppended)
+            }
+          } else {
+            // Keep append-only implementation for efficiency.
```

Comment:
Let's remove `Keep`:

`// Append-only implementation of getting appended files for efficiency.`

Also, can you add comment that in `else`, there should be no deleted files, which is checked in `isHybridScanCandidate`? I think a better way is to keep a context to pass around. (we should do this as a follow up).